lib/active_merchant/billing/gateways/visanet_peru.rb

```diff
@@ -0,0 +1,239 @@
+        requires!(options, :login, :password)
+        super
+      end
+
+      def purchase(amount, payment_method, options={})
+        MultiResponse.run() do |r|
+          r.process { authorize(amount, payment_method, options) }
+          r.process { capture(r.authorization, options) }
+        end
+      end
+
+      def authorize(amount, payment_method, options={})
+        #JSON
+        post = {}
+
+        add_invoice(post, amount, options)
+        add_payment_method(post, payment_method)
+        add_customer_data(post, options)
+        add_device_fingerprint_data(post, options)
+        add_merchant_define_data(post, options)
+
+        # No vaulting for now
+        post[:createAlias] = false
+
+        commit("authorize", post, options)
+      end
+
+      def capture(authorization, options={})
+        params = {}
+        action, merchant_id, purchase_number = split_authorization(authorization)
+        options[:merchant_id] = merchant_id
+        options[:purchaseNumber] = purchase_number
+        params[:externalTransactionId] = purchase_number
+        commit("capture", params, options)
+      end
+
+      def void(authorization, options={})
+        params = {}
+        action, merchant_id, purchase_number = split_authorization(authorization)
+        options[:merchant_id] = merchant_id
+        options[:purchaseNumber] = purchase_number
+        params[:externalTransactionId] = purchase_number
+        puts options
+        case action
+        when "authorize"
+          commit("void", params, options)
+        when "capture"
+          commit("cancel", params, options)
+        end
+      end
+
+      # def refund(amount, authorization, options={})
+      # end
+
+      # def credit(amount, payment_method, options={})
+      # end
+
+      def verify(credit_card, options={})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      # def store(payment_method, options = {})
+      #   post = {}
+      #   add_payment_method(post, payment_method)
+      #   add_customer_data(post, options)
+
+      #   commit("store", post)
+      # end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        #JSON.
+        transcript.
+          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
+          gsub(%r((\"cardNumber\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cvv2Code\\\":)\d+), '\1[FILTERED]')
+      end
+
+      private
+
+      CURRENCY_CODES = Hash.new{|h,k| raise ArgumentError.new("Unsupported currency: #{k}")}
+      CURRENCY_CODES["USD"] = 840
+      CURRENCY_CODES["PEN"] = 604
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money).to_f
+        # Visanet Peru expects a 9-digit numeric purchaseNumber
+        post[:purchaseNumber] = options[:order_id]
```

Comment:
Let's go ahead and do this for the purchaseNumber:

```
post[:purchaseNumber] = options[:purchase_number] || rand(999999999)
```

Since it's such a specific format that should probably get truncated, we shouldn't try to conflate it with the "real" order id. Let's give it its own option and generate a default one if not present?